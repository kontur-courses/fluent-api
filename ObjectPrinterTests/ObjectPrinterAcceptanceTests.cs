using System.Collections;
using System.Globalization;
using FluentAssertions;
using ObjectPrinting;
using ObjectPrinting.Tests;

namespace ObjectPrinterTests;

[TestFixture]
public class ObjectPrinterAcceptanceTests
{
    private readonly string newLine = Environment.NewLine;

    [Test]
    public void ObjectPrinter_OnSubsequentPrints_ShouldNotContainCycle()
    {
        var person = new Person { Name = "Alex", Age = 19 };

        var printer = ObjectPrinter.For<Person>();
        var result = printer.PrintToString(person);
        var result2 = printer.PrintToString(person);
        result2.Should().NotContain("Cycle");
    }

    [Test]
    public void ObjectPrinter_WhenExcludingIntType_ShouldPrintObjectWithoutIntPropertiesAndFields()
    {
        var person = new Person { Name = "Alex", Age = 19 };

        var printer = ObjectPrinter.For<Person>();
        var result = printer.Excluding<int>().PrintToString(person);
        result.Should().NotContain(nameof(person.Age));
        result.Should().NotContain(nameof(person.Defects));
    }

    [Test]
    public void ObjectPrinter_WhenExcludingMember_ShouldPrintObjectWithoutMember()
    {
        var person = new Person { Name = "Alex", Age = 19 };
        var printer = ObjectPrinter.For<Person>();
        var result = printer.Excluding(x => x.Age).PrintToString(person);
        result.Should().NotContain(nameof(person.Age));
    }

    [Test]
    public void ObjectPrinter_WhenPrintingPropertyUsingSomeConditional_ShouldPrintObjectWithModifiedProperty()
    {
        var person = new Person { Name = "Alex", Age = 15 };
        const string nameProperty = nameof(person.Age);
        var printer = ObjectPrinter.For<Person>();
        var result = printer.Printing(x => x.Age).Using(x => nameProperty).PrintToString(person);
        result.Should().Contain($"{nameProperty} = {nameProperty}");
    }

    [Test]
    public void ObjectPrinter_WhenChangingPropertiesByType_ShouldBeObjectWithModifiedProperty()
    {
        var person = new Person { Name = "Alex", Age = 15, Height = 1.2 };

        var printer = ObjectPrinter.For<Person>();
        var result = printer.Printing<double>().Using(x => (x * x).ToString()).PrintToString(person);
        result.Should().Contain($"{nameof(person.Height)} = {person.Height * person.Height}");
    }

    [Test]
    public void PrintToString_WhenPrintAllMembersInObject_ShouldPrintAllPublicMembers()
    {
        var person = new Person { Name = "Alex", Age = 15, Height = 1.2 };
        var printer = ObjectPrinter.For<Person>();
        var result = printer.PrintToString(person);
        foreach (var property in person.GetType().GetProperties())
        {
            result.Should().Contain($"{property.Name}");
        }

        foreach (var field in person.GetType().GetFields())
        {
            result.Should().Contain($"{field.Name}");
        }
    }

    [Test]
    public void ObjectPrinter_WhenCultureIsSet_ShouldPrintPropertyWithCulture()
    {
        var person = new Person { Name = "Alex", Age = 15, Height = 2.4 };
        var culture = new CultureInfo("en-GB");
        var printer = ObjectPrinter.For<Person>();
        var result = printer.Printing(x => x.Defects).Using(culture).PrintToString(person);
        result.Should().Contain($"{nameof(person.Defects)} = {person.Defects.ToString(culture)}");
    }

    [Test]
    public void ObjectPrinter_WhenTrimmedStringProperties_ShouldPrintObjectWithTrimmedProperties()
    {
        const int maxLen = 1;
        var person = new Person { Name = "Alex", Age = 15, Height = 2.4, Id = Guid.Empty };
        var printer = ObjectPrinter.For<Person>();
         printer.Printing<string>().TrimmedToLength(maxLen);
        var result = person.PrintToString();
        result.Should().Contain($"{nameof(person.Name)} = {person.Name[..maxLen]}");
    }

    [Test]
    public void ObjectPrinter_WhenTrimmedStringPropertiesButCroppingLengthLess0_ShouldTrowArgumentException()
    {
        var person = new Person { Name = "Alex", Age = 15, Height = 2.4, Id = Guid.Empty };
        var printer = ObjectPrinter.For<Person>();
        var action = () => { printer.Printing<string>().TrimmedToLength(-10).PrintToString(person); };
        action.Should().Throw<ArgumentException>("Error: The length of the truncated string cannot be negative");
    }

    [Test]
    public void ObjectPrinter_WhenPropertyRefersItself_ShouldPrintObjectWithCycleProperty()
    {
        var kid = new Kid { Name = "Pasha" };
        var parent = new Kid { Name = "Lev" };
        kid.Parent = parent;
        parent.Parent = kid;


        var printer = ObjectPrinter.For<Kid>();
        var result = printer.PrintToString(kid);
        result.Should().Contain(kid.GetType().Name);
         result.Should().Contain($"\t{nameof(kid.Parent)} = {kid.GetType().Name}");
        result.Should().Contain($"\t\t{nameof(kid.Parent)} = (Cycle){kid.Parent.GetType().FullName}");
    }

    [Test]
    public void ObjectPrinter_WhenPrintingDictionaryProperty_ShouldPrintObject()
    {
        var dictionary = new Dictionary<int, string>
        {
            { 1, "hello" },
            { 2, "hell" },
            { 3, "hel" },
            { 4, "he" },
            { 5, "h" }
        };
        var collections = new Collections();
        collections.Dictionary = new Dictionary<int, string>(dictionary);
        var printer = ObjectPrinter.For<Collections>();
        var result = printer.PrintToString(collections);
        result.Should().Contain($"{nameof(collections.Dictionary)}");
        foreach (var value in dictionary) result.Should().Contain($"{value.Key}{newLine} : {value.Value}{newLine}");
    }
    [Test]
    public void ObjectPrinter_WhenPrintingIEnumerableProperty_ShouldPrintObject()
    {
        var collections = new Collections();
        collections.Enumerable = new Stack<int>(new[] { 1, 2, 3, 4 });
        var printer = ObjectPrinter.For<Collections>();
        var result = printer.PrintToString(collections);
        foreach (var value in collections.Enumerable) result.Should().Contain($"{value}{newLine}");
    }

    [Test]
    public void ObjectPrinter_WhenThereIsObjectWithListProperty_ShouldPrintObject()
    {
        var collections = new Collections();
        collections.List = new List<object> { 1, 2, 3 };
        var printer = ObjectPrinter.For<Collections>();
        var result = printer.PrintToString(collections);
        foreach (var value in collections.List) result.Should().Contain($"{value}{newLine}");
    }

    [Test]
    public void ObjectPrinter_WhenThereIsArrayGenericObjects_ShouldPrintObject()
    {
        var collections = new Collections();
        collections.Array = new[] { new[] { 1, 2, 3 } };
        var printer = ObjectPrinter.For<Collections>();
        var result = printer.PrintToString(collections);
        result.Should().Contain($"{nameof(collections.Array)} = {newLine}");
        foreach (var list in collections.Array)
        {
            result.Should().Contain($"\t\t{list.GetType().Name}{newLine}");
            foreach (var value in list) result.Should().Contain($"\t\t\t{value}{newLine}");
        }
    }

    [Test]
    public void ObjectPrinter_WhenThereIsEnumerableTypeRefersItself_ShouldPrintObject()
    {
        var collections = new Collections();
        collections.List = new List<object>();
        collections.List.Add(collections.List);
        var printer = ObjectPrinter.For<Collections>();
        var result = printer.PrintToString(collections);
        result.Should().Contain($"{nameof(collections.List)} = {newLine}");
        foreach (var list in collections.List) result.Should().Contain($"\t\t\t(Cycle){list.GetType().FullName}");
    }

    [Test]
    public void ObjectPrinter_WhenPrintingSomeClassesInList_ShouldPrintObject()
    {
        var collections = new Collections();
        var child = new Person { Name = "Child" };
        collections.Persons = new List<Person> { new() { Name = "Lev" }, child, child };
        var printer = ObjectPrinter.For<Collections>();
        var result = printer.PrintToString(collections);
        result.Should().Contain($"{nameof(collections.Persons)} = ");
        foreach (var person in collections.Persons)
        {
            result.Should().Contain($"\t\t{person.GetType().Name}");
            result.Should().Contain($"\t\t\t{nameof(person.Name)} = {person.Name}{newLine}");
            result.Should().Contain($"\t\t\t{nameof(person.Age)} = {person.Age}{newLine}");
        }
    }
    [Test]
    public void ObjectPrinter_WhenPrintingIdenticalPersons_ShouldPrintObjectWithoutCycle()
    {
        var person1 = new Kid() { Name = "Abobus" };
        person1.Parent = new Kid() { Name = "Abobus" };
        var result = person1.PrintToString();
        result.Should().NotContain("(Cycle)");
    }
}