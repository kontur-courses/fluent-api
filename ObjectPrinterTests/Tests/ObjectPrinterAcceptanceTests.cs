using System;
using System.Globalization;
using ApprovalTests;
using ApprovalTests.Reporters;
using NUnit.Framework;
using ObjectPrinterTests;

namespace ObjectPrinting.Solved.Tests
{
    [TestFixture]
    public class ObjectPrinterAcceptanceTests
    {
        // 1
        [UseReporter(typeof(DiffReporter))]
        [Test]
        public void ObjectPrinter_Can_ExcludeTypeInt()
        {
            var person = MockPerson.GetCoolProgramer;

            var printer = ObjectPrinter.For<Person>().Excluding<int>();
            Approvals.Verify(printer.PrintToString(person));
        }

        [UseReporter(typeof(DiffReporter))]
        [Test] // 6
        public void ObjectPrinter_Can_PropertyNameExclude()
        {
            var person = MockPerson.GetCoolProgramer;

            var printer = ObjectPrinter.For<Person>().Excluding(p => p.Name);
            Approvals.Verify(printer.PrintToString(person));
        }

        [UseReporter(typeof(DiffReporter))]
        [Test] //2 
        public void ObjectPrinter_Can_SerializeForTypeInt()
        {
            var person = MockPerson.GetCoolProgramer;

            var printer = ObjectPrinter.For<Person>().Printing<double>()
                .Using(p => $"какой-то double {p}");

            Approvals.Verify(printer.PrintToString(person));
        }

        [UseReporter(typeof(DiffReporter))]
        [Test] // 4
        public void ObjectPrinter_Can_SerializeProperty()
        {
            var person = MockPerson.GetCoolProgramer;

            var printer = ObjectPrinter.For<Person>()
                .Printing(p => p.Name).Using(name => $"этого челика зовут {name}");

            Approvals.Verify(printer.PrintToString(person));
        }

        [UseReporter(typeof(DiffReporter))]
        [Test] // 3
        public void ObjectPrinter_Can_ChangeCulture()
        {
            var person = MockPerson.GetCoolProgramer;

            var printer = ObjectPrinter.For<Person>()
                .Printing<double>().Using(new CultureInfo("ru-RU"));

            Approvals.Verify(printer.PrintToString(person));
        }

        [UseReporter(typeof(DiffReporter))]
        [Test] // 5
        public void ObjectPrinter_Can_TrimString()
        {
            var person = MockPerson.GetCoolProgramer;

            var printer = ObjectPrinter.For<Person>()
                .Printing(p => p.Name).TrimmedToLength(7);

            Approvals.Verify(printer.PrintToString(person));
        }


        [UseReporter(typeof(DiffReporter))]
        [Test]
        public void ObjectPrinter_Can_ArraySerialize()
        {
            var house = Room.GetStudioHouse();

            var printer = ObjectPrinter.For<Room>();
            
            Approvals.Verify(printer.PrintToString(house));
        }
        
        
        [UseReporter(typeof(DiffReporter))]
        [Test]
        public void ObjectPrinter_Can_WorkWithCombination()
        {
            var person = new Person { Name = "Alex", Age = 19, Height = 12.1};

        
            var printer = ObjectPrinter.For<Person>()
                .Excluding<Guid>()
                .Printing<double>().Using(new CultureInfo("ru-RU"))
                .Printing(p => p.Name).Using(x => $"новое имя челика {x}")
                .Printing(p => p.Name).TrimmedToLength(2)
                .Excluding(p => p.Age);
        
            var s1 = printer.PrintToString(person);
            Approvals.Verify(s1);
        }

        [UseReporter(typeof(DiffReporter))]
        [Test]
        public void ObjectPrinter_Can_WorkWithArray()
        {
            var shop = Shop.GetShop();

            var s = shop.PrintToString(x => x.Printing(s => s.NameShop).Using(x => $"ваще прикольный магаз {x}"));
            
            Approvals.Verify(s);
        }
        
        [UseReporter(typeof(DiffReporter))]
        [Test]
        public void ObjectPrinter_Can_WithPrimitiveTypes()
        {
            var shop = TestTypePropetries.GetTestType();

            var s = shop.PrintToString();
            
            Approvals.Verify(s);
        }


        [UseReporter(typeof(DiffReporter))]
        [Test]
        public void ObjectPrinter_Can_WorkWithDict()
        {
            var cookies = CookiesWithData.GetFakeCookiesWithData();

            var s = cookies.PrintToString();

            Approvals.Verify(s);
        }
    }
}