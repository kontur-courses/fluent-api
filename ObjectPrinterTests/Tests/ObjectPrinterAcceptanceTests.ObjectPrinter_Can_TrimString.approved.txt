Person
	Id = 00000000-0000-0000-0000-000000000000
	Name = Андрей 
	Height = 19.2
	Age = 20
	PlaceBirth = Uknown
